docs/proposals/lifecycle_events.md

```diff
@@ -0,0 +1,128 @@
+# Support execution of the commands based on lifecycle events
+
+## Abstract
+Devfile 2.0.0 allows binding of commands to specific lifecycle events: https://github.com/devfile/kubernetes-api/issues/32. These commands are meant to be executed when those specific lifecycle events are triggered.
+
+The lifecycle events that have been implemented in Devfile 2.0.0 are generic. Their effect/meaning might be slightly different in the context a workspace (Che) or a single project (odo). However, odo should aim to provide a similar user experience within the scope of a single project.
+
+## Motivation
+Devfiles support commands that can be triggered based on development lifecycle events. With the implementation of lifecycle events, stack creators will be able to leverage all the capabilities of devfiles when building the correct experience/features for their stacks. `odo` should support/execute these commands at appropriate times within `odo` development flow.
+
+## User Stories
+Each lifecycle event would be suited to individual user stories. We propose that at least one issue is made for each of them for tracking and implementation purposes.
+
+- Add support for `postStart` commands: [container initialization support](https://github.com/openshift/odo/issues/2936)
+- Add support for `preStop` commands: 
+
+## Design overview
+Some of the events might not be useful for `odo` to adopt. We should only support the ones that have a clear use-case, and then add more as other use-cases emerge.
+
+For odo, the most important events are the `postStart`, and the `preStop` because they have clear, reasonable use case within odo’s flow. There is no clear use case within odo for `preStart` and `postStop` today.
+
+- `preStart` - Do not support in odo
+- `postStart` - support executing these commands as part of component initialization
+- `preStop` - support executing these commands as part of the component delete
+- `postStop` - Do not support in odo
```

Comment:
I think that this can be supported in odo. It can simply execute after `preStop` commands.

Reply:
We can support `postStart`, but it has no real value as it will be same as `preStop`. We have listed this as a futures as it might be better to support the ones where there is a clear use case and then build up.

@kadel are you recommending that we support both and do the same thing for them effectively? I feel  that might be more confusing for devfile creators.

Reply:
> We have listed this as a futures as it might be better to support the ones where there is a clear use case and then build up.

It already got into devfile spec, so someone probably has a use case for it. 


> @kadel are you recommending that we support both and do the same thing for them effectively? It might be ok for completeness but I feel that might be more confusing for devfile creators.

I don't see any problems with this.
But I see a problem when there is devfile that uses this event and someone tries to use it with odo. Would we error out and say that this devfile is unsupported? Or just show a warning and never execute the commands? Both are pretty bad options :-(

Reply:
It's not clear on how we want to implement this. If someone specify a `postStop` command using the build/run container, we'll be starting up the exact containers that we just stopped.

Reply:
I see your point on partial implementation. However, if we consider stop is equivalent to deleting the component and containers, the flow would be:
```
- preStop
- stop (delete component and all containers)
- postStop
```

The issue now is that by the time we have done the stop/delete operation, we don't have any existing deployment where we can execute the the `postStop` commands, and I don't think we want to create another one and then delete it for this.

If I change this to what you are proposing:

```
- preStop
- postStop
- stop (delete component and all containers)
```

Then we really are changing the meaning of postStop and I would be concerned about that. 

I feel the real use case for odo could potentially come from actions that need to be take on "local" target. As in you want to clean up some temporary files from the developer's workspace on the host. That could be a good fit, but the way to execute them will be different. than running them on the cluster, and I'm not sure if odo supports that today.

Implementing this will be trivial as we just need to parse few more commands, but I'm trying to get clarity on use case before we go down this path.

Reply:
> It's not clear on how we want to implement this. If someone specify a `postStop` command using the build/run container, we'll be starting up the exact containers that we just stopped.

yes, and if they define it like this it is what they wanted.

Reply:
> I see your point on partial implementation. However, if we consider stop is equivalent to deleting the component and containers, the flow would be:
> 
> ```
> - preStop
> - stop (delete component and all containers)
> - postStop
> ```

That is true. This can be fairly easily implemented. We can just create standalone Pod or Job after the component is deleted.

Reply:
I've tried to re-read : https://github.com/devfile/kubernetes-api/issues/32 and found something that I had missed earlier: 

`Implementation of preStart commands may need execution of a workspace pod init container. That means that the component associated to the preStart command should not be part of the workspace pod. The same applies to postStop bindings.`

There is a clear requirement that container used for preStart and postStart commands will not overlap with containers of the main workspace pod. This makes is a bit clearer and perhaps we can implement postStop as: 

- create a new deployment for running just these postStop command
- run the postStop command
- delete the deployment

or change the meaning slightly and:
- run the commands in the same component after preStop but before deleting the component and all containers

Reply:
What would we be deleting with `PostStop` when starting a new Job or Pod in the cluster? a PVC? a database? In these scenarios, we don't need to create a resource instead we would be deleting a resource. Example: a logging sidecar receiving information from the runtime container. Then it makes sense to delete the runtime container, before deleteing the logging sidecar.
Even though, I still think this step makes more sense in a Che workspace, and maybe in the local docker support were we have create volumes locally or configuration files on the host.

Reply:
I'm think these commands will more be like stop the database and back up data to a volume. vs simply deleting a container.

Reply:
We can update the proposal to use a standalone pod for postStart commands and then delete the pod. all of this will be done as part of `odo delete`.

I'll reach out to the devfiles team to get more clarity on the use case they had in mind too.